Implement failglob

pdebelak opened the pull request:
This implements failglob for osh. I'm not sure about the error message setup here, it could be too verbose, but bash prints a message for this in addition to setting a non-zero status. Also could be missing some contexts, but this appears to do the right thing for command context, loop context, and array literals.

In reference to #1036

andychu commented:
Amazing, thank you!  You even changed the documentation :)

How about making the error message 

    'Pattern %r matched no files (failglob)'  % arg

to remind the user that they have failglob on.

Or maybe:

    'failglob: Pattern %r matched no files'  % arg

Also, one thing I noticed is that despite passing the  spec tests, this behaves slightly differently than bash.  But I think that is OK!

The issue is a semicolon vs. newline, which I think is extremely obscure.

Compare

```
shopt -s failglob
echo *.zz; echo status=$?  # bash doesn't execute the second part !!!

echo *.zz
echo status=$?  # bash executes this

```

I just added this to: https://github.com/oilshell/oil/wiki/Shell-WTFs

I noticed this in some other cases -- occasionally the `;` is significant.  But OSH in general doesn't respect it, and I don't think we should start.

I think it will probably mess up the code to try to respect it, whereas now the logic is clean.

But bash is the only shell that implements failglob.  So it might be nice to add a test case and mark OSH as OK.  Just so we have a record of the difference.

------

I think the error message is more important though ... the `;` is really trivia.  I believe it's due to bash constructing a fairly naive parse tree and then putting the stack unwinding in a specific spot, not a really intentional design decision.

pdebelak commented:
I've updated the error message to include the pattern and use the `failglob` prefix so it reads like `failglob: Pattern '*.ZZ' matched no files` and added a failing spec test for the bizarre `;` behavior. This should be ready for re-review.

Also, fun note, I keep typing `failblog` instead of `failglob`.

andychu commented:
Great thank you! 

I followed up with a commit to implement it for extended globs, which are separate from regular globs since they're more static (happen before word splitting).

I hope to do a release with this great work by the end of the week, or maybe next week!
